File: go/sdk/snapshot_package.go
package sdk

import (
	"archive/tar"
	"bytes"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"os"
	pathpkg "path"
	"path/filepath"
	"strconv"
	"strings"
	"sync"
)

// CompletedAlarmSnapshotPackage 是通过完整性与安全校验后的告警抓拍图像包。
type CompletedAlarmSnapshotPackage struct {
	EventID       uint64
	PackageSHA256 string
	PackagePath   string
}

type snapshotPackageState struct {
	packageLength uint64
	packageSHA256 [32]byte
	chunkCount    uint32
	received      map[uint32]struct{}
}

// AlarmSnapshotPackageAssembler 将类型 2 分块有界、磁盘优先地重组为 USTAR 图像包。
// 同一实例可并发调用；只有完整包、receipt.json 和目录元数据均持久化后才返回完成结果。
type AlarmSnapshotPackageAssembler struct {
	outputDir        string
	maxPendingEvents int
	mu               sync.Mutex
	events           map[uint64]*snapshotPackageState
}

func NewAlarmSnapshotPackageAssembler(outputDir string, maxPendingEvents int) (*AlarmSnapshotPackageAssembler, error) {
	if strings.TrimSpace(outputDir) == "" {
		return nil, errors.New("outputDir 不能为空")
	}
	if maxPendingEvents < 1 {
		return nil, errors.New("maxPendingEvents 必须大于 0")
	}
	if err := os.MkdirAll(outputDir, 0o700); err != nil {
		return nil, fmt.Errorf("创建抓拍图像目录失败: %w", err)
	}
	return &AlarmSnapshotPackageAssembler{
		outputDir: outputDir, maxPendingEvents: maxPendingEvents,
		events: make(map[uint64]*snapshotPackageState),
	}, nil
}

func (a *AlarmSnapshotPackageAssembler) Accept(chunk *EvidencePackageChunk) (*CompletedAlarmSnapshotPackage, error) {
	if err := validateSnapshotChunk(chunk); err != nil {
		return nil, err
	}
	a.mu.Lock()
	defer a.mu.Unlock()

	eventDir := filepath.Join(a.outputDir, strconv.FormatUint(chunk.EventID, 10))
	if err := os.MkdirAll(eventDir, 0o700); err != nil {
		return nil, fmt.Errorf("创建事件目录失败: %w", err)
	}
	hashHex := hex.EncodeToString(chunk.PackageSHA256[:])
	partPath := filepath.Join(eventDir, hashHex+".tar.part")
	finalPath := filepath.Join(eventDir, hashHex+".tar")

	if completed, err := completedSnapshotPackage(finalPath, eventDir, chunk, hashHex); err != nil {
		return nil, err
	} else if completed != nil {
		delete(a.events, chunk.EventID)
		return completed, nil
	}

	state := a.events[chunk.EventID]
	if state == nil {
		if len(a.events) >= a.maxPendingEvents {
			return nil, errors.New("待接收告警抓拍图像事件数量超过有界限制")
		}
		state = &snapshotPackageState{
			packageLength: chunk.PackageLength,
			packageSHA256: chunk.PackageSHA256,
			chunkCount:    chunk.ChunkCount,
			received:      make(map[uint32]struct{}),
		}
		a.events[chunk.EventID] = state
	}
	if state.packageLength != chunk.PackageLength || state.packageSHA256 != chunk.PackageSHA256 || state.chunkCount != chunk.ChunkCount {
		return nil, errors.New("同一 eventId 的抓拍图像包标识冲突")
	}

	file, err := os.OpenFile(partPath, os.O_CREATE|os.O_RDWR, 0o600)
	if err != nil {
		return nil, fmt.Errorf("打开抓拍图像临时包失败: %w", err)
	}
	defer file.Close()
	if _, duplicate := state.received[chunk.ChunkIndex]; duplicate {
		actual := make([]byte, len(chunk.Chunk))
		if _, err := file.ReadAt(actual, int64(chunk.ChunkOffset)); err != nil && err != io.EOF {
			return nil, fmt.Errorf("读取重复分块失败: %w", err)
		}
		if !bytes.Equal(actual, chunk.Chunk) {
			return nil, errors.New("重复分块的字节内容冲突")
		}
	} else {
		if _, err := file.WriteAt(chunk.Chunk, int64(chunk.ChunkOffset)); err != nil {
			return nil, fmt.Errorf("写入抓拍图像分块失败: %w", err)
		}
		if err := file.Sync(); err != nil {
			return nil, fmt.Errorf("持久化抓拍图像分块失败: %w", err)
		}
		state.received[chunk.ChunkIndex] = struct{}{}
	}
	if len(state.received) != int(state.chunkCount) {
		return nil, nil
	}
	if err := file.Close(); err != nil {
		return nil, err
	}

	if err := validateCompleteSnapshotPackage(partPath, state.packageLength, state.packageSHA256); err != nil {
		delete(a.events, chunk.EventID)
		_ = os.Remove(partPath)
		return nil, err
	}
	if err := os.Rename(partPath, finalPath); err != nil {
		return nil, fmt.Errorf("提交抓拍图像包失败: %w", err)
	}
	if err := writeSnapshotReceipt(eventDir, chunk.EventID, hashHex, filepath.Base(finalPath)); err != nil {
		return nil, err
	}
	if err := syncDirectory(eventDir); err != nil {
		return nil, err
	}
	delete(a.events, chunk.EventID)
	return &CompletedAlarmSnapshotPackage{EventID: chunk.EventID, PackageSHA256: hashHex, PackagePath: finalPath}, nil
}

func validateSnapshotChunk(chunk *EvidencePackageChunk) error {
	if chunk == nil {
		return errors.New("抓拍图像分块不能为空")
	}
	if chunk.MessageType != 2 || chunk.HeaderLength != evidencePackageHeaderLength || chunk.PackageFormat != 1 || chunk.EvidenceKind != 1 {
		return errors.New("抓拍图像分块格式不受支持")
	}
	if chunk.EventID == 0 || chunk.PackageLength == 0 || chunk.PackageLength > maxEvidencePackageBytes {
		return errors.New("抓拍图像包身份或长度非法")
	}
	expectedCount := uint32((chunk.PackageLength + evidenceChunkBytes - 1) / evidenceChunkBytes)
	expectedOffset := uint64(chunk.ChunkIndex) * evidenceChunkBytes
	if chunk.ChunkCount != expectedCount || chunk.ChunkIndex >= chunk.ChunkCount || chunk.ChunkOffset != expectedOffset || expectedOffset >= chunk.PackageLength {
		return errors.New("抓拍图像包分块范围非法")
	}
	expectedLength := min(uint64(evidenceChunkBytes), chunk.PackageLength-expectedOffset)
	if uint64(len(chunk.Chunk)) != expectedLength {
		return errors.New("抓拍图像包分块长度非法")
	}
	if chunk.ChunkIndex == 0 && (len(chunk.Chunk) < 262 || string(chunk.Chunk[257:262]) != "ustar") {
		return errors.New("抓拍图像包首块缺少 USTAR 标识")
	}
	return nil
}

func completedSnapshotPackage(finalPath, eventDir string, chunk *EvidencePackageChunk, hashHex string) (*CompletedAlarmSnapshotPackage, error) {
	info, err := os.Stat(finalPath)
	if errors.Is(err, os.ErrNotExist) {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}
	if uint64(info.Size()) != chunk.PackageLength {
		return nil, errors.New("已完成抓拍图像包长度冲突")
	}
	file, err := os.Open(finalPath)
	if err != nil {
		return nil, err
	}
	actual := make([]byte, len(chunk.Chunk))
	_, readErr := file.ReadAt(actual, int64(chunk.ChunkOffset))
	closeErr := file.Close()
	if readErr != nil && readErr != io.EOF {
		return nil, readErr
	}
	if closeErr != nil {
		return nil, closeErr
	}
	if !bytes.Equal(actual, chunk.Chunk) {
		return nil, errors.New("已完成抓拍图像包的重复分块冲突")
	}
	if err := validateCompleteSnapshotPackage(finalPath, chunk.PackageLength, chunk.PackageSHA256); err != nil {
		return nil, err
	}
	if err := writeSnapshotReceipt(eventDir, chunk.EventID, hashHex, filepath.Base(finalPath)); err != nil {
		return nil, err
	}
	if err := syncDirectory(eventDir); err != nil {
		return nil, err
	}
	return &CompletedAlarmSnapshotPackage{EventID: chunk.EventID, PackageSHA256: hashHex, PackagePath: finalPath}, nil
}

func validateCompleteSnapshotPackage(packagePath string, expectedLength uint64, expectedHash [32]byte) error {
	info, err := os.Stat(packagePath)
	if err != nil {
		return err
	}
	if uint64(info.Size()) != expectedLength {
		return errors.New("重组后的抓拍图像包长度不匹配")
	}
	file, err := os.Open(packagePath)
	if err != nil {
		return err
	}
	hash := sha256.New()
	if _, err := io.Copy(hash, file); err != nil {
		file.Close()
		return err
	}
	if err := file.Close(); err != nil {
		return err
	}
	if !bytes.Equal(hash.Sum(nil), expectedHash[:]) {
		return errors.New("重组后的抓拍图像包 SHA-256 不匹配")
	}
	return validateSafeUSTAR(packagePath)
}

func validateSafeUSTAR(packagePath string) error {
	file, err := os.Open(packagePath)
	if err != nil {
		return err
	}
	defer file.Close()
	reader := tar.NewReader(file)
	first := true
	seen := make(map[string]struct{})
	for {
		header, err := reader.Next()
		if errors.Is(err, io.EOF) {
			break
		}
		if err != nil {
			return fmt.Errorf("抓拍图像 USTAR 非法: %w", err)
		}
		name := header.Name
		if first && name != "manifest.json" {
			return errors.New("抓拍图像 USTAR 的第一项必须是 manifest.json")
		}
		first = false
		if !safeUSTARName(name) || (header.Typeflag != tar.TypeReg && header.Typeflag != tar.TypeRegA) {
			return errors.New("抓拍图像 USTAR 包含不安全或非普通文件成员")
		}
		if _, duplicate := seen[name]; duplicate {
			return errors.New("抓拍图像 USTAR 包含重复成员")
		}
		seen[name] = struct{}{}
	}
	if first {
		return errors.New("抓拍图像 USTAR 不能为空")
	}
	return nil
}

func safeUSTARName(name string) bool {
	if name == "" || strings.Contains(name, "\\") || pathpkg.IsAbs(name) {
		return false
	}
	clean := pathpkg.Clean(name)
	return clean == name && clean != "." && clean != ".." && !strings.HasPrefix(clean, "../")
}

func writeSnapshotReceipt(eventDir string, eventID uint64, hashHex, packageName string) error {
	receipt := map[string]string{
		"eventId": strconv.FormatUint(eventID, 10), "kind": "SNAPSHOT",
		"packageSha256": hashHex, "package": packageName,
	}
	payload, err := json.Marshal(receipt)
	if err != nil {
		return err
	}
	temporaryPath := filepath.Join(eventDir, "receipt.json.tmp")
	receiptPath := filepath.Join(eventDir, "receipt.json")
	file, err := os.OpenFile(temporaryPath, os.O_CREATE|os.O_TRUNC|os.O_WRONLY, 0o600)
	if err != nil {
		return err
	}
	if _, err = file.Write(payload); err == nil {
		err = file.Sync()
	}
	if closeErr := file.Close(); err == nil {
		err = closeErr
	}
	if err != nil {
		return err
	}
	if err := os.Rename(temporaryPath, receiptPath); err != nil {
		return err
	}
	return nil
}

func syncDirectory(directory string) error {
	file, err := os.Open(directory)
	if err != nil {
		return err
	}
	defer file.Close()
	return file.Sync()
}

File: javascript/sdk/snapshot-package.js
"use strict";

const crypto = require("node:crypto");
const fs = require("node:fs");
const path = require("node:path");

const HEADER_LENGTH = 76;
const CHUNK_BYTES = 128 * 1024;
const MAX_PACKAGE_BYTES = 32 * 1024 * 1024;
const TAR_BLOCK_BYTES = 512;

class AlarmSnapshotPackageAssembler {
  constructor(outputDirectory, options = {}) {
    if (!outputDirectory) {
      throw new Error("outputDirectory 不能为空");
    }
    this.maxPendingEvents = options.maxPendingEvents ?? 8;
    if (!Number.isInteger(this.maxPendingEvents) || this.maxPendingEvents < 1) {
      throw new Error("maxPendingEvents 必须大于 0");
    }
    this.outputDirectory = path.resolve(outputDirectory);
    fs.mkdirSync(this.outputDirectory, { recursive: true, mode: 0o700 });
    this.events = new Map();
  }

  accept(chunk) {
    validateChunk(chunk);
    const eventId = String(chunk.eventId);
    const eventDirectory = path.join(this.outputDirectory, eventId);
    fs.mkdirSync(eventDirectory, { recursive: true, mode: 0o700 });
    const hashHex = String(chunk.packageSha256);
    const partialPath = path.join(eventDirectory, `${hashHex}.tar.part`);
    const finalPath = path.join(eventDirectory, `${hashHex}.tar`);

    if (fs.existsSync(finalPath)) {
      validateCompletePackage(finalPath, Number(chunk.packageLength), hashHex);
      const actual = readFileRange(finalPath, Number(chunk.chunkOffset), chunk.chunk.length);
      if (!actual.equals(Buffer.from(chunk.chunk))) {
        throw new Error("已完成抓拍图像包的重复分块冲突");
      }
      writeReceipt(eventDirectory, eventId, hashHex, path.basename(finalPath));
      syncDirectory(eventDirectory);
      this.events.delete(eventId);
      return { eventId, packageSha256: hashHex, packagePath: finalPath };
    }

    let state = this.events.get(eventId);
    if (!state) {
      if (this.events.size >= this.maxPendingEvents) {
        throw new Error("待接收告警抓拍图像事件数量超过有界限制");
      }
      state = {
        packageLength: String(chunk.packageLength),
        packageSha256: hashHex,
        chunkCount: chunk.chunkCount,
        received: new Set(),
      };
      this.events.set(eventId, state);
    }
    if (state.packageLength !== String(chunk.packageLength)
        || state.packageSha256 !== hashHex || state.chunkCount !== chunk.chunkCount) {
      throw new Error("同一 eventId 的抓拍图像包标识冲突");
    }

    const data = Buffer.from(chunk.chunk);
    const descriptor = fs.openSync(partialPath, fs.existsSync(partialPath) ? "r+" : "w+", 0o600);
    try {
      if (state.received.has(chunk.chunkIndex)) {
        const actual = readExact(descriptor, Number(chunk.chunkOffset), data.length);
        if (!actual.equals(data)) {
          throw new Error("重复分块的字节内容冲突");
        }
      } else {
        writeExact(descriptor, Number(chunk.chunkOffset), data);
        fs.fsyncSync(descriptor);
        state.received.add(chunk.chunkIndex);
      }
    } finally {
      fs.closeSync(descriptor);
    }

    if (state.received.size !== state.chunkCount) {
      return null;
    }
    try {
      validateCompletePackage(partialPath, Number(chunk.packageLength), hashHex);
    } catch (error) {
      this.events.delete(eventId);
      fs.rmSync(partialPath, { force: true });
      throw error;
    }
    fs.renameSync(partialPath, finalPath);
    writeReceipt(eventDirectory, eventId, hashHex, path.basename(finalPath));
    syncDirectory(eventDirectory);
    this.events.delete(eventId);
    return { eventId, packageSha256: hashHex, packagePath: finalPath };
  }
}

function validateChunk(chunk) {
  if (!chunk || chunk.messageType !== 2 || chunk.headerLength !== HEADER_LENGTH
      || chunk.packageFormat !== 1 || chunk.evidenceKind !== 1) {
    throw new Error("抓拍图像分块格式不受支持");
  }
  if (!/^[1-9][0-9]*$/.test(String(chunk.eventId))
      || !/^[0-9a-f]{64}$/.test(String(chunk.packageSha256))) {
    throw new Error("抓拍图像包身份非法");
  }
  const packageLength = parseBoundedInteger(chunk.packageLength, "packageLength");
  const chunkOffset = parseBoundedInteger(chunk.chunkOffset, "chunkOffset");
  if (packageLength < 1 || packageLength > MAX_PACKAGE_BYTES) {
    throw new Error("抓拍图像包长度非法");
  }
  const expectedCount = Math.ceil(packageLength / CHUNK_BYTES);
  const expectedOffset = chunk.chunkIndex * CHUNK_BYTES;
  if (!Number.isInteger(chunk.chunkIndex) || !Number.isInteger(chunk.chunkCount)
      || chunk.chunkCount !== expectedCount || chunk.chunkIndex < 0
      || chunk.chunkIndex >= chunk.chunkCount || chunkOffset !== expectedOffset
      || expectedOffset >= packageLength) {
    throw new Error("抓拍图像包分块范围非法");
  }
  const expectedLength = Math.min(CHUNK_BYTES, packageLength - expectedOffset);
  const data = Buffer.from(chunk.chunk);
  if (data.length !== expectedLength) {
    throw new Error("抓拍图像包分块长度非法");
  }
  if (chunk.chunkIndex === 0
      && (data.length < 262 || data.subarray(257, 262).toString("ascii") !== "ustar")) {
    throw new Error("抓拍图像包首块缺少 USTAR 标识");
  }
}

function parseBoundedInteger(value, name) {
  const text = String(value);
  if (!/^(0|[1-9][0-9]*)$/.test(text)) {
    throw new Error(`${name} 不是非负整数`);
  }
  const result = Number(text);
  if (!Number.isSafeInteger(result)) {
    throw new Error(`${name} 超出 JavaScript 安全整数范围`);
  }
  return result;
}

function validateCompletePackage(packagePath, expectedLength, expectedHash) {
  const info = fs.statSync(packagePath);
  if (!info.isFile() || info.size !== expectedLength) {
    throw new Error("重组后的抓拍图像包长度不匹配");
  }
  const digest = crypto.createHash("sha256");
  const descriptor = fs.openSync(packagePath, "r");
  try {
    const buffer = Buffer.allocUnsafe(64 * 1024);
    for (let position = 0; position < expectedLength;) {
      const count = fs.readSync(descriptor, buffer, 0, Math.min(buffer.length, expectedLength - position), position);
      if (count === 0) throw new Error("抓拍图像包提前结束");
      digest.update(buffer.subarray(0, count));
      position += count;
    }
  } finally {
    fs.closeSync(descriptor);
  }
  if (digest.digest("hex") !== expectedHash) {
    throw new Error("重组后的抓拍图像包 SHA-256 不匹配");
  }
  validateSafeUstar(packagePath, expectedLength);
}

function validateSafeUstar(packagePath, packageLength) {
  const descriptor = fs.openSync(packagePath, "r");
  let first = true;
  let terminated = false;
  const names = new Set();
  try {
    for (let offset = 0; offset + TAR_BLOCK_BYTES <= packageLength;) {
      const header = readExact(descriptor, offset, TAR_BLOCK_BYTES);
      if (header.every((value) => value === 0)) {
        if (offset + 2 * TAR_BLOCK_BYTES > packageLength
            || !readExact(descriptor, offset + TAR_BLOCK_BYTES, TAR_BLOCK_BYTES)
              .every((value) => value === 0)) {
          throw new Error("抓拍图像 USTAR 结束块非法");
        }
        terminated = true;
        break;
      }
      if (header.subarray(257, 262).toString("ascii") !== "ustar" || !validTarChecksum(header)) {
        throw new Error("抓拍图像 USTAR Header 非法");
      }
      const basicName = nullTerminatedAscii(header, 0, 100);
      const prefix = nullTerminatedAscii(header, 345, 155);
      const name = prefix ? `${prefix}/${basicName}` : basicName;
      if (first && name !== "manifest.json") {
        throw new Error("抓拍图像 USTAR 的第一项必须是 manifest.json");
      }
      first = false;
      const type = header[156];
      if (!safeTarName(name) || (type !== 0 && type !== 0x30)) {
        throw new Error("抓拍图像 USTAR 包含不安全或非普通文件成员");
      }
      if (names.has(name)) {
        throw new Error("抓拍图像 USTAR 包含重复成员");
      }
      names.add(name);
      const size = parseTarOctal(header, 124, 12);
      const next = offset + TAR_BLOCK_BYTES + Math.ceil(size / TAR_BLOCK_BYTES) * TAR_BLOCK_BYTES;
      if (!Number.isSafeInteger(next) || next > packageLength) {
        throw new Error("抓拍图像 USTAR 成员长度非法");
      }
      offset = next;
    }
  } finally {
    fs.closeSync(descriptor);
  }
  if (first || !terminated) {
    throw new Error("抓拍图像 USTAR 为空或未正确结束");
  }
}

function validTarChecksum(header) {
  const expected = parseTarOctal(header, 148, 8);
  let actual = 0;
  for (let index = 0; index < header.length; index += 1) {
    actual += index >= 148 && index < 156 ? 32 : header[index];
  }
  return expected === actual;
}

function parseTarOctal(header, offset, length) {
  const value = header.subarray(offset, offset + length).toString("ascii").replaceAll("\0", "").trim();
  if (!value) return 0;
  if (!/^[0-7]+$/.test(value)) throw new Error("USTAR 八进制字段非法");
  const result = Number.parseInt(value, 8);
  if (!Number.isSafeInteger(result)) throw new Error("USTAR 数值字段过大");
  return result;
}

function nullTerminatedAscii(header, offset, length) {
  const field = header.subarray(offset, offset + length);
  const end = field.indexOf(0);
  return field.subarray(0, end < 0 ? field.length : end).toString("ascii");
}

function safeTarName(name) {
  if (!name || name.startsWith("/") || name.includes("\\")) return false;
  return name.split("/").every((part) => part && part !== "." && part !== "..");
}

function readFileRange(filePath, offset, length) {
  const descriptor = fs.openSync(filePath, "r");
  try {
    return readExact(descriptor, offset, length);
  } finally {
    fs.closeSync(descriptor);
  }
}

function readExact(descriptor, offset, length) {
  const result = Buffer.alloc(length);
  for (let total = 0; total < length;) {
    const count = fs.readSync(descriptor, result, total, length - total, offset + total);
    if (count === 0) throw new Error("抓拍图像包提前结束");
    total += count;
  }
  return result;
}

function writeExact(descriptor, offset, data) {
  for (let total = 0; total < data.length;) {
    total += fs.writeSync(descriptor, data, total, data.length - total, offset + total);
  }
}

function writeReceipt(eventDirectory, eventId, hashHex, packageName) {
  const receiptPath = path.join(eventDirectory, "receipt.json");
  const temporaryPath = `${receiptPath}.tmp`;
  const payload = Buffer.from(JSON.stringify({
    eventId, kind: "SNAPSHOT", packageSha256: hashHex, package: packageName,
  }));
  const descriptor = fs.openSync(temporaryPath, "w", 0o600);
  try {
    writeExact(descriptor, 0, payload);
    fs.fsyncSync(descriptor);
  } finally {
    fs.closeSync(descriptor);
  }
  fs.renameSync(temporaryPath, receiptPath);
}

function syncDirectory(directory) {
  const descriptor = fs.openSync(directory, "r");
  try {
    fs.fsyncSync(descriptor);
  } finally {
    fs.closeSync(descriptor);
  }
}

module.exports = { AlarmSnapshotPackageAssembler };
